add loopback_test file.

naveedjamil opened the pull request.

Commit:
Add Loopback_test File

This increases the test coverage for the package pkg/loopback. Initially there were no test files present in this module.
Include a flag for pkg/loopback in test-unit file.

Signed-off-by: Naveed Jamil <naveed.jamil@tenpearl.com>

aaronlehmann commented on pkg/loopback/loopback_test.go:
```diff
@@ -0,0 +1,100 @@
+	"github.com/stretchr/testify/assert"
+)
+
+const fileName = "/var/test/foo.bar"
```
I think it's not a good practice to use a fixed file name here. It means, for example, that running multiple instances of the test at once will fail. Generally tests use `ioutil.TempFile` to create a temporary file, then delete that file with a deferred `os.Remove` call.

aaronlehmann commented on pkg/loopback/loopback_test.go:
```diff
@@ -0,0 +1,100 @@
+	size := int64(10 * 1024 * 1024)
+	fd, err := os.Create(fileName)
+	if err != nil {
+		return errors.New("Error creating file")
```
If you import `github.com/pkg/errors`, you can `return errors.Wrap(err, "error creating file")`. This will preserve the original error but add some context to it.

aaronlehmann commented on pkg/loopback/loopback_test.go:
```diff
@@ -0,0 +1,100 @@
+	assert.Equal(t, ErrAttachLoopbackDevice, err)
+
+	err = createSparseFile(fileName)
+	assert.Nil(t, err, "Error creating sparse file")
```
Use `assert.NoError(t, err, "error creating sparse file")`.

aaronlehmann commented on pkg/loopback/loopback_test.go:
```diff
@@ -0,0 +1,100 @@
+	assert.Nil(t, err, "Error creating sparse file")
+	defer os.Remove(fileName)
+
+	loopDevice, err := os.Open("/dev/loop5")
```
If a test requires root privileges, I think it's good practice to require a flag to run it, so it does not break `go test <packagename>` for non-root users (and also so that the test does not have unexpected side effects on the system).

There's a good example of this in containerd; see the `init` and `RequiresRoot` functions in https://github.com/containerd/containerd/blob/master/testutil/helpers.go. Tests that need root privileges call `RequiresRoot(t)` before doing anything else.

naveedjamil replied:
In the `hack/make/test-unit` file test cases of all packages present in moby/moby are executed with the same parameters so providing a default parameter causes all of the other functions to throw an error `flag provided but not defined: [flag name]`. So either we will have to conditionally check whether test cases of this package are being executed or implement the flag in all of the packages which would be pointless. So what would you suggest we do in this case ?

aaronlehmann replied:
I think either of those alternatives would be fine. You could either split testing of this package into a separate `go test` command in the shell script (and add other packages to that command as they gain support for this option in the future), or you could define the flag in the `init` function of a package that is imported by all packages which are tested.

For now the former approach is probably easiest.

Commit:
Add test coverage to pkg/signal

Signed-off-by: Naveed Jamil <naveed.jamil@tenpearls.com>

Commit:
Add test coverage to pkg/term/proxy.go

Signed-off-by: Naveed Jamil <naveed.jamil@tenpearls.com>

runcom committed:
daemon: correctly try to retrieve init/runtime versions

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

Commit:
Add a error check in postHijacked to avoid docker exec command blocking.

When user execute docker exec command, docker daemon maybe have err return because
of ExecExists check, and then the hijack stream will not be close, it can lead to
docker exec command block.

Signed-off-by: yangshukui <yangshukui@huawei.com>

Commit:
Logging driver should receive same file in start/stop request

Signed-off-by: Peter Bücker <peter.buecker@gmail.com>

thaJeztah committed:
Merge pull request #33583 from naag/33569-logging-plugin-file-inconsistency

Pass the same log filename to logging plugins on start/stop requests

vieux committed:
Merge pull request #33546 from runcom/init-runtime-info

daemon: correctly try to retrieve init/runtime versions

crosbymichael committed:
Merge pull request #33520 from naveed-jamil-tenpearls/pkg/term

Add test coverage to pkg/term

Commit:
daemon: Remove daemon datastructure dump functionality

When sending SIGUSR1 to the daemon, it can crash because of a concurrent
map access panic, showing a stack trace involving dumpDaemon. It appears
it's not possible to recover from a concurrent map access panic. Since
it's important that SIGUSR1 not be a destructive operation, sadly the
best course of action I can think of is to remove this functionality.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

vieux committed:
Merge pull request #33451 from naveed-jamil-tenpearls/pkg/signal/signal

Add test coverage to pkg/signal

coolljt0725 committed:
Merge pull request #33262 from keloyang/exec-block

Add a error check in postHijacked to avoid docker exec command blocking.

cpuguy83 committed:
Merge pull request #33598 from aaronlehmann/remove-dumpdaemon

daemon: Remove daemon datastructure dump functionality
